File: client/ios/PastePoint/Core/Services/Moderation/BlockService.swift
//
//  Copyright © 2026 PastePoint. All rights reserved.
//  SPDX-License-Identifier: GPL-3.0-only
//

import Combine
import Foundation

@MainActor
final class BlockService: ObservableObject {

  @Published private(set) var blockedPeers: Set<String> = []

  private var cancellables = Set<AnyCancellable>()

  init(wsService: WebSocketConnectionService) {
    wsService.$isConnected
      .receive(on: DispatchQueue.main)
      .sink { [weak self] connected in
        if !connected { self?.clear() }
      }
      .store(in: &cancellables)
  }

  func block(_ peer: String) {
    guard !peer.isEmpty else { return }

    log.info("blocking peer")
    blockedPeers.insert(peer)
  }

  func unblock(_ peer: String) {
    log.info("unblocking peer")
    blockedPeers.remove(peer)
  }

  func isBlocked(_ peer: String) -> Bool {
    blockedPeers.contains(peer)
  }

  private func clear() {
    guard !blockedPeers.isEmpty else { return }

    log.debug("clearing blocks — identities are reassigned on reconnect")
    blockedPeers.removeAll()
  }
}

File: client/web/src/app/core/services/communication/block.service.ts
import { Injectable, inject } from '@angular/core';
import { BehaviorSubject } from 'rxjs';
import { NGXLogger } from 'ngx-logger';
import { WebSocketConnectionService } from './websocket-connection.service';

@Injectable({
  providedIn: 'root',
})
export class BlockService {
  private wsService = inject(WebSocketConnectionService);
  private logger = inject(NGXLogger);

  public blockedPeers$ = new BehaviorSubject<Set<string>>(new Set());
  private blockedPeers = new Set<string>();

  constructor() {
    this.wsService.connected$.subscribe(() => this.clear());
  }

  public block(peer: string): void {
    if (!peer) {
      return;
    }

    this.logger.info('block', `Blocking peer ${peer}`);
    this.blockedPeers.add(peer);
    this.blockedPeers$.next(new Set(this.blockedPeers));
  }

  public unblock(peer: string): void {
    this.logger.info('unblock', `Unblocking peer ${peer}`);

    this.blockedPeers.delete(peer);
    this.blockedPeers$.next(new Set(this.blockedPeers));
  }

  public isBlocked(peer: string): boolean {
    return this.blockedPeers.has(peer);
  }

  private clear(): void {
    if (this.blockedPeers.size === 0) {
      return;
    }

    this.logger.debug('clear', 'Clearing blocks — identities are reassigned on reconnect');
    this.blockedPeers.clear();
    this.blockedPeers$.next(new Set());
  }
}
